File: Data Structures & Algorithms/permutation-string/submission-3.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int m = s1.length();
        if (m > s2.length()) return false;

        int[] target = new int[26];
        for (char c : s1.toCharArray()) {
            target[c - 'a']++;
        }

        int[]window = new int[26];
        int left = 0;
        for(int right = 0; right < s2.length();right++){
            int rc = s2.charAt(right) - 'a';
            window[rc]++;
            while(window[rc] > target[rc]){
                window[s2.charAt(left)- 'a']--;
                left++;
            }
            if (right - left + 1 == m) return true;

        }
        return false;
    }
}

File: Data Structures & Algorithms/minimum-size-subarray-sum/submission-3.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        res = len(nums) + 1
        s = 0
        left = 0
        for right, n in enumerate(nums):
            s += nums[right]
            while s >= target:
                res = min(res, right - left + 1)
                s -= nums[left]
                left += 1
        return res if res <= len(nums) else 0
